[AArch64] Add option -msve-streaming-vector-bits= .

efriedma-quic opened the pull request:
This is similar to -msve-vector-bits, but for streaming mode: it constrains the legal values of "vscale", allowing optimizations based on that constraint.

This also fixes conversions between SVE vectors and fixed-width vectors in streaming functions with -msve-vector-bits and -msve-streaming-vector-bits.

This rejects any use of arm_sve_vector_bits types in streaming functions; if it becomes relevant, we could add arm_sve_streaming_vector_bits types in the future.

This doesn't touch the __ARM_FEATURE_SVE_BITS define.

efriedma-quic committed:
[AArch64] Add option -msve-streaming-vector-bits= .

This is similar to -msve-vector-bits, but for streaming mode: it
constrains the legal values of "vscale", allowing optimizations based on
that constraint.

This also fixes conversions between SVE vectors and fixed-width vectors
in streaming functions with -msve-vector-bits and
-msve-streaming-vector-bits.

This currently doesn't touch the __ARM_FEATURE_SVE_BITS define or the
arm_sve_vector_bits attribute.

efriedma-quic committed:
Fix interaction with streaming-compatible functions.

efriedma-quic commented:
Pushed revised patch which correctly handles streaming-compatible functions (the previous version accidentally treated them as non-streaming).

efriedma-quic commented on clang/include/clang/Basic/TargetInfo.h:
```diff
@@ -1034,9 +1034,16 @@ class TargetInfo : public TransferrableTargetInfo,
   virtual std::optional<std::pair<unsigned, unsigned>>
-  getVScaleRange(const LangOptions &LangOpts, bool IsArmStreamingFunction,
+  getVScaleRange(const LangOptions &LangOpts,
+                 ArmStreamingKind IsArmStreamingFunction,
```
LangOpts is global.  IsArmStreamingFunction is context-dependent: it depends on whether we're currently analyzing a function marked __arm_streaming/etc.

We could reorganize the code, maybe... like, move some bits of the implementation into AST.  But I'm not sure that really helps.

wangpc-pp replied:
Okay I didn't how it is used. Please go ahead.

sdesmalen-arm commented on clang/lib/Basic/Targets/AArch64.cpp:
```diff
@@ -820,16 +820,23 @@ AArch64TargetInfo::getTargetBuiltins() const {
+    return std::pair<unsigned, unsigned>(
+        LangOpts.VScaleStreamingMin ? LangOpts.VScaleStreamingMin : 1,
+        LangOpts.VScaleStreamingMax);
+
```
When compiling a `__arm_streaming_compatible` function with
`-march=armv9-a+sme -Xclang -mvscale-min=2 -Xclang -mvscale-streaming-min=2  -Xclang -mvscale-max=4 -Xclang -mvscale-streaming-max=4`, it sets the vscale_range to `vscale_range(1, 16)`. I think it should set this to `vscale_range(2, 4)` instead.

sdesmalen-arm commented:
> This currently doesn't touch the __ARM_FEATURE_SVE_BITS define or the arm_sve_vector_bits attribute

We actually need to be a bit more careful to guard against code that uses the `arm_sve_vector_bits` types in streaming[-compatible] contexts when the streaming and non-streaming vector lengths are not the same.

For example:
```
typedef svint32_t int32x16_t __attribute__((arm_sve_vector_bits(512)));

int32x16_t foo_512(int32x16_t a, int32x16_t b) __arm_streaming {
  return a + b;
}
```

Compiling this with `-msve-vector-bits=512 -msve-streaming-vector-bits=256`, Clang will create `foo_512` with `vscale_range(2, 2)`, and to extract the <16 x i32> value from the scalable function argument, it will generate:
```call <16 x i32> @llvm.vector.extract.v16i32.nxv4i32(<vscale x 4 x i32> %a.coerce, i64 0)```

This is effectively undefined if vscale is not >= 4.

efriedma-quic committed:
Address review comments.

efriedma-quic committed:
Fix formatting

efriedma-quic committed:
Merge branch 'main' into msve-streaming-vector-bits

MacDue commented on clang/lib/Basic/Targets/AArch64.cpp:
```diff
@@ -820,16 +820,23 @@ AArch64TargetInfo::getTargetBuiltins() const {
 std::optional<std::pair<unsigned, unsigned>>
 AArch64TargetInfo::getVScaleRange(const LangOptions &LangOpts,
-                                  bool IsArmStreamingFunction,
+                                  ArmStreamingKind IsArmStreamingFunction,
```
I think calling the enum `ArmStreamingMode` and the parameter `StreamingMode` would be okay ("`StreamingKind`" is a little inconsistent as everywhere else this is referred to as the streaming mode).

efriedma-quic committed:
Merge branch 'main' into msve-streaming-vector-bits

efriedma-quic commented on clang/lib/Sema/Sema.cpp:
```diff
@@ -2261,6 +2261,21 @@ void Sema::checkTypeSupport(QualType Ty, SourceLocation Loc, ValueDecl *D) {
+      } else if (const auto *FTy = FD->getType()->getAs<FunctionProtoType>()) {
+        if (FTy->getAArch64SMEAttributes() &
+            FunctionType::SME_PStateSMCompatibleMask) {
+          Diag(Loc, diag::err_sve_fixed_vector_in_streaming_function)
```
I can allow the use of arm_sve_vector_bits in streaming/streaming-compatible functions if the streaming and non-streaming widths are the same, sure.

Your last two testcases are the same: the conversion happens in the caller, so the first case reduces to the second.  Which is an issue, I guess, but not directly related to this patch.  It doesn't involve any fixed-width types.

efriedma-quic committed:
Address review comments.

Fix variable names, allow fixed-width types in streaming functions if
the streaming and non-streaming widths are the same.

efriedma-quic committed:
Fix build.

efriedma-quic commented:
Ping

efriedma-quic committed:
Apply suggestions from code review

Co-authored-by: Sander de Smalen <sander.desmalen@arm.com>

efriedma-quic committed:
Add relnote

efriedma-quic committed:
Fix formatting.
